pyomo/network/port.py

```diff
@@ -0,0 +1,606 @@
+                          else None
+
+        self.vars = {}
+        self._arcs = []
+        self._sources = []
+        self._dests = []
+        self._rules = {}
+        self._splitfracs = ComponentMap()
+
+    def __getstate__(self):
+        state = super(_PortData, self).__getstate__()
+        for i in _PortData.__slots__:
+            state[i] = getattr(self, i)
+        return state
+
+    # Note: None of the slots on this class need to be edited, so we
+    # don't need to implement a specialized __setstate__ method, and
+    # can quietly rely on the super() class's implementation.
+
+    def __getattr__(self, name):
+        """Returns self.vars[name] if it exists"""
+        if name in self.vars:
+            return self.vars[name]
+        # Since the base classes don't support getattr, we can just
+        # throw the "normal" AttributeError
+        raise AttributeError("'%s' object has no attribute '%s'"
+                             % (self.__class__.__name__, name))
+
+    def arcs(self, active=None):
+        """A list of Arcs in which this Port is a member"""
+        if active is None:
+            return list(self._arcs)
+        tmp = []
+        for a in self._arcs:
+            if a.active == active:
+                tmp.append(a)
+        return tmp
+
+    def sources(self, active=None):
+        """A list of Arcs in which this Port is a destination"""
+        if active is None:
+            return list(self._sources)
+        tmp = []
+        for a in self._sources:
+            if a.active == active:
+                tmp.append(a)
+        return tmp
+
+    def dests(self, active=None):
+        """A list of Arcs in which this Port is a source"""
+        if active is None:
+            return list(self._dests)
+        tmp = []
+        for a in self._dests:
+            if a.active == active:
+                tmp.append(a)
+        return tmp
+
+    def set_value(self, value):
+        """Cannot specify the value of a port"""
+        raise ValueError("Cannot specify the value of a port: '%s'" % self.name)
+
+    def polynomial_degree(self):
+        ans = 0
+        for v in self._iter_vars():
+            tmp = v.polynomial_degree()
+            if tmp is None:
+                return None
+            ans = max(ans, tmp)
+        return ans
+
+    def is_fixed(self):
+        """Return True if all vars/expressions in the Port are fixed"""
+        return all(v.is_fixed() for v in self._iter_vars())
+
+    def is_potentially_variable(self):
+        """Return True as ports may (should!) contain variables"""
+        return True
+
+    def is_binary(self):
+        return len(self) and all(v.is_binary() for v in self._iter_vars())
+
+    def is_integer(self):
+        return len(self) and all(v.is_integer() for v in self._iter_vars())
+
+    def is_continuous(self):
+        return len(self) and all(v.is_continuous() for v in self._iter_vars())
```

Comment:
Not all of them fail, it seems the only method that expressions don't have is `is_continuous`. Example:

```
>>> from pyomo.environ import *
>>> from pyomo.network import *
>>> m = ConcreteModel()
>>> m.p = Port()
>>> m.x = Var()
>>> m.p.add(m.x)
>>> m.p.add(-m.x, "foo")
>>> m.p.is_binary()
False
>>> m.p.is_integer()
False
```

What should be done about `is_continuous`? Could we just remove the method? To be honest I found all these as-is in Connector so I'm not too attached to any of them so if you don't see a point for them I suppose we could remove any number of them.

Reply:
Whoops, I didn't realize it was short circuiting once it found that x was not binary or integer. When you add `-m.x` first you get the attribute error for all of these methods. I'll change it so that it identifies all the variables in expressions.